ocaml/lambda/value_rec_compiler.ml

```diff
@@ -0,0 +1,837 @@
+        (* New primitives -- need to check *)
+        Misc.fatal_error "size_of_primitives: unhandled primitive"
+  in
+  compute_expression_size Ident.Map.empty lam
+
+let lfunction_with_body { kind; params; return; body = _; attr; loc;
+                          mode; ret_mode; region } body =
+  lfunction' ~kind ~params ~return ~body ~attr ~loc ~mode ~ret_mode ~region
+
+(** {1. Function Lifting} *)
+
+(* The compiler allows recursive definitions of functions that are not
+   syntactic functions:
+   {[
+     let rec f_syntactic_function = fun x ->
+       f_syntactic_function x
+
+     let rec g_needs_lift =
+       let () = ... in
+       (fun x -> g_needs_lift (foo x))
+
+     let rec h_needs_lift_and_closure =
+       let v = ref 0 in
+       (fun x -> incr v; h_needs_lift_and_closure (bar x))
+
+     let rec i_needs_lift_and_eta =
+       let aux x = i_needs_lift_and_eta (baz x) in
+       aux
+   ]}
+
+   We need to translate those using only syntactic functions or blocks.
+   For some functions, we only need to lift a syntactic function in tail
+   position from its surrounding context:
+   {[
+     let rec g_context =
+       let () = ... in
+       ()
+     and g_lifted = fun x ->
+       g_lifted (foo x)
+   ]}
+
+   In general the function may refer to local variables, so we perform
+   a local closure conversion before lifting:
+   {[
+     let rec h_context =
+       let v = ref 0 in
+       { v }
+     and h_lifted = fun x ->
+       incr h_context.v;
+       h_lifted (bar x)
+   ]}
+   Note that the closure environment computed from the context is passed as a
+   mutually recursive definition, that is, a free variable, and not as an
+   additional function parameter (which is customary for closure conversion).
+
+   Finally, when the tail expression is a variable, we perform an eta-expansion
+   to get a syntactic function, that we can then close and lift:
+   {[
+     let rec i_context =
+       let aux x = i_lifted (baz x) in
+       { aux }
+     and i_lifted = fun x -> i_context.aux x
+   ]}
+*)
+
+type lifted_function =
+  { lfun : Lambda.lfunction;
+    free_vars_block_size : int;
+  }
+
+type 'a split_result =
+  | Unreachable
+  | Reachable of lifted_function * 'a
+
+let ( let+ ) res f =
+  match res with
+  | Unreachable -> Unreachable
+  | Reachable (func, lam) -> Reachable (func, f lam)
+
+(* The closure blocks are immutable.
+   (Note: It is usually safe to declare immutable blocks as mutable,
+   but in this case the blocks might be empty and declaring them as Mutable
+   would cause errors later.) *)
+let lifted_block_mut : Lambda.mutable_flag = Immutable
+let lifted_block_read_sem : Lambda.field_read_semantics = Reads_agree
+
+let no_loc = Debuginfo.Scoped_location.Loc_unknown
+
+let rec split_static_function block_var local_idents lam :
+  Lambda.lambda split_result =
+  match lam with
+  | Lvar _v ->
+    (* CR vlaviron: We cannot eta-expand when arbitrary layouts are involved. *)
```

Comment:
The note in comments is the key. Upstream we can generate an eta-expansion without knowing the arity, because everything is a value, but we could generate better code if we propagated the arity.
Here we can't generate the eta-expansion without knowing the arity (including layouts), so I had to implement the arity propagation (it's in commit f2a8efca23158caf216d25df15beb0f574b891c5).
I guess we could push this upstream too, since it's done anyway.

Reply:
I was reviewing commit-by-commit and forgot to delete this comment before posting. It's all fine as is.